### DISCUSSION TITLE
Cant Get It To Work

### DISCUSSION THREAD

**Discussion by Deadtrix21** [mikke89/RmlUi#256] [Q&A] [answered]
Hello...
I have been struggling for a few hours and i just cant get this thing to setup

Can someone hop into discord with me and show me please cause i am at a long big bloody loss 

or can someone hop into discord and show me how to set it up 
IDE : VS 2022

**Comment by mikke89** [comment, marked as answer]
Hi.

Although, I can't help you on Discord, if you describe your problem here I'll try to help. It might be useful for others too.

Are you trying to build the samples, or integrate it into your project? What have you done so far?

**Reply by Deadtrix21** [reply]
So what I am trying to do is 
Is I wanna try and actually use it..... But I have no clue where to start what

Like I wanted to use this one because of the CSS and HTML which is really nice because u can model them to pixel perfection 

But this one I don't understand how to add it to VS2019 at all
Or even how to use it

I am on windows so yeah
Right now I am using nuklear
Which is c and cpp GUI

And it actually works
It's all built in a single header too

Nor do I know how to compile the samples 😅

Sorry I know I am walking trouble

**Reply by Deadtrix21** [reply]
I am still studying 
Cpp and C and D3X9

I know how to use C#
JS and some other languages

**Reply by mikke89** [reply]
Alright, I see where you're coming from, we all started here so no worries. C++ is notoriously fragmented when it comes to including libraries. It sounds like you've only integrated header-only libraries so far. Those are the easy ones, for everything else there are a thousand different ways to include a library and link it up. I would really encourage you to learn the basics for linking libraries. It isn't necessarily that difficult, but it's quite a steep learning curve at first and the fragmentation doesn't help.

Start with the basics of CMake, at least for building other libraries including RmlUi. For RmlUi this is explained in detail [in the documentation](https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/building_with_cmake.html), see the section for building on Windows and make sure to build the samples and check them out too. Once you have built the library you need to link it up to your own program, and include the header files. This is also [explained in detail](https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/integrating.html) for Visual Studio so that should be a perfect match for you. 

There are also package managers like vcpkg and Conan that may be helpful, but I'd encourage you to learn the manual steps first. Unfortunately, the package managers aren't quite mature enough yet that we don't have to think about the underlying behavior.

Once you have RmlUi all linked up, the next step would be to write the interfaces. Most of the work will be writing the render interface for your DIrectX backend. Again, this is really only about reading the documentation in detail, and then drawing the triangles RmlUi hands you. There are some other DirectX render interfaces for RmlUi around that you could take a look at for inspiration.

Best of luck, just need to put your head down and get going, the documentation should be your friend all the way. I'm sure you will have learned a ton once you have things up and working :)

**Reply by Deadtrix21** [reply]
Can u recommend any books on where to learn or see this in action as well as how hears work

**Reply by mikke89** [reply]
You could have a look at Stroustrup's "Programming: Principles and Practice Using C++", it's a good introduction to C++ and has some basic information about linking. For learning CMake I'd recommend taking a look at this introduction to modern CMake: https://cliutils.gitlab.io/modern-cmake/.
